File: week3-exercises02-scala/src/test/scala/com/datastax/workshop/Ex09_Query4c_ReadMetrics.scala
package com.datastax.workshop

import java.util.UUID

import org.junit.jupiter.api._
import org.junit.platform.runner.JUnitPlatform
import org.junit.runner.RunWith

import com.datastax.oss.driver.api.core.cql.Row

@RunWith(classOf[JUnitPlatform])
object Ex09_Query4c_ReadMetrics extends ExerciseBase("Exercise4")

@RunWith(classOf[JUnitPlatform])
class Ex09_Query4c_ReadMetrics {

  import Ex09_Query4c_ReadMetrics._

  @Test
  def read_measurements(): Unit = {

    LOGGER.info("----- Speed Measurements -----")
    journeyRepo
      .findSpeedMeasurements(SPACECRAFT, UUID.fromString(JOURNEY_ID))
      .zipWithIndex
      .foreach { case row -> index => logMeasurement("speed", index, row) }

    LOGGER.info("----- Temperature Measurements -----")
    journeyRepo
      .findTemperatureMeasurements(SPACECRAFT, UUID.fromString(JOURNEY_ID))
      .zipWithIndex
      .foreach { case row -> index => logMeasurement("temperature", index, row) }

    LOGGER.info("----- Pressure Measurements -----")
    journeyRepo
      .findPressureMeasurements(SPACECRAFT, UUID.fromString(JOURNEY_ID))
      .zipWithIndex
      .foreach { case row -> index => logMeasurement("pressure", index, row) }

    LOGGER.info("----- Location Measurements -----")
    journeyRepo
      .findLocationMeasurements(SPACECRAFT, UUID.fromString(JOURNEY_ID))
      .zipWithIndex
      .foreach { case row -> index => logLocationMeasurement(index, row) }

    LOGGER.info("SUCCESS")
  }

  def logMeasurement(kind: String, index: Int, row: Row): Unit =
    LOGGER.info(
      "idx:{}, time={}, value={}",
      index,
      row.getInstant("reading_time"),
      row.getDouble(kind)
    )

  def logLocationMeasurement(index: Int, row: Row): Unit = {
    LOGGER.info(
      "idx:{}, time={}, location(x={}, y={}, z={})",
      index,
      row.getInstant("reading_time"),
      row.getUdtValue("location").getDouble("x_coordinate"),
      row.getUdtValue("location").getDouble("y_coordinate"),
      row.getUdtValue("location").getDouble("z_coordinate")
    )
  }
}

File: week3-exercises01-java/src/test/java/com/datastax/workshop/Ex09_Query4c_ReadMetrics.java
package com.datastax.workshop;

import java.util.UUID;
import java.util.List;

import org.junit.jupiter.api.Test;
import org.junit.platform.runner.JUnitPlatform;
import org.junit.runner.RunWith;

import com.datastax.oss.driver.api.core.cql.ResultSet;
import com.datastax.oss.driver.api.core.cql.Row;
import com.datastax.oss.driver.api.core.cql.SimpleStatement;

@RunWith(JUnitPlatform.class)
public class Ex09_Query4c_ReadMetrics extends ExerciseBase {

    public Ex09_Query4c_ReadMetrics() {
        super("Exercise4");
    }

    @Test
    public void read_measurements() {

        LOGGER.info("----- Speed Measurements -----");
        List<Row> rows = journeyRepo.findSpeedMeasurements(SPACECRAFT, UUID.fromString(JOURNEY_ID));
        int offset = 0;
        for (Row row : rows) {
            logMeasurement("speed", ++offset, row);
        }

        LOGGER.info("----- Temperature Measurements -----");
        rows = journeyRepo.findTemperatureMeasurements(SPACECRAFT, UUID.fromString(JOURNEY_ID));
        offset = 0;
        for (Row row : rows) {
            logMeasurement("temperature", ++offset, row);
        }

        LOGGER.info("----- Pressure Measurements -----");
        rows = journeyRepo.findPressureMeasurements(SPACECRAFT, UUID.fromString(JOURNEY_ID));
        offset = 0;
        for (Row row : rows) {
            logMeasurement("pressure", ++offset, row);
        }

        LOGGER.info("----- Location Measurements -----");
        rows = journeyRepo.findLocationMeasurements(SPACECRAFT, UUID.fromString(JOURNEY_ID));
        offset = 0;
        for (Row row : rows) {
            logLocationMeasurement(++offset, row);
        }

        LOGGER.info("SUCCESS");
    }

    private void logMeasurement(String kind, int index, Row row) {
        LOGGER.info("idx:{}, time={}, value={}", index, row.getInstant("reading_time"), row.getDouble(kind));
    }

    private void logLocationMeasurement(int index, Row row) {
        LOGGER.info("idx:{}, time={}, location(x={}, y={}, z={})", index, row.getInstant("reading_time"),
                row.getUdtValue("location").getDouble("x_coordinate"),
                row.getUdtValue("location").getDouble("y_coordinate"),
                row.getUdtValue("location").getDouble("z_coordinate"));
    }
}
